qiskit_experiments/curve_analysis/drawer_mixin.py

```diff
@@ -0,0 +1,512 @@
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Curve drawing functionality for analysis."""
+
+import functools
+from typing import Callable, Set, List, Dict, Sequence, Union, Optional
+
+import numpy as np
+from matplotlib.axes import Axes
+from matplotlib.ticker import FuncFormatter, ScalarFormatter
+from qiskit.utils import detach_prefix
+from uncertainties import unumpy as unp, UFloat
+
+from qiskit_experiments.curve_analysis.curve_data import FitData
+from qiskit_experiments.framework import AnalysisResultData, Options
+from qiskit_experiments.framework.matplotlib import get_non_gui_ax
+
+
+class CurveDrawerMixin:
+    """Mixin to provide drawing functionalities for curve fit result."""
+
+    # Internal state for drawing options
+    _draw_options: Options
+    _set_draw_options: Set
+
+    def __init_subclass__(cls, **kwargs):
+        """Initialize drawing options."""
+        super().__init_subclass__(**kwargs)
+
+        # This is a hack to let the mixin have internal state for drawing option.
+        @functools.wraps(cls.__init__, assigned=("__annotations__",))
+        def __new__(cls, *args, **kwargs):
+            instance = super().__new__(cls, *args, **kwargs)
+            setattr(instance, "_default_draw_options", cls._default_draw_options)
+            setattr(instance, "_set_draw_options", set())
+            return instance
+
+        cls.__new__ = __new__
+
+    def _initialize_canvas(self) -> Axes:
+        """Initialize matplotlib axis.
+
+        Curve analysis drawer supports 2D sub-graph visualization.
+        This first checks the drawing options if matplotlib axis is provided by users.
+        Axis is initialized only when axis is not provided.
+        Initialized axis is set to the drawing options.
+
+        Returns:
+            Handler to matplotlib axis object. This may be initialized with inset axes.
+        """
+        # Create axis if empty
+        if not self.draw_options.axis:
+            axis = get_non_gui_ax()
+
+            # Set figure size if axis is not provided.
+            # Otherwise user may set the figure size by oneself.
+            # Then avoid applying draw defaults to avoid override user's preference.
+            figure = axis.get_figure()
+            figure.set_size_inches(*self.draw_options.figsize)
+        else:
+            axis = self.draw_options.axis
+
+        n_rows, n_cols = self.draw_options.subplots
+        n_subplots = n_cols * n_rows
+        if n_subplots > 1:
+            # Add inset axis. User may provide a single axis object via the analysis option,
+            # while this analysis tries to draw its result in multiple canvases,
+            # especially when the analysis consists of multiple curves.
+            # Inset axis is experimental implementation of matplotlib 3.0 so maybe unstable API.
```

Comment:
Is this a good idea if the API is unstable?

Reply:
This already exists in the current codebase and this is unavoidable. I investigated the Matplotlib but there is no approach other than this to add inset axes. Since this is not user-facing API in our codebase, we can update the code when something is deprecated in Matplotlib.